First draft of a .NET code generator handling calc-base-of-base, calc-base and calc-lib

assyadh opened the pull request:
This is the first draft of a .NET code generator handling calc-base-of-base, calc-base and calc-lib

The generator is split in 3 files:

- dotnet.ts : the target class containing the bulk of the generator code.

- dotnet/filegenerator.ts: a file generator class which generates miscellaneous .NET required files (the csproj and assembly info file)

- dotnet/nameutils.ts: Various naming utility classes

This generator produces a 1:1 output with the calc-base-of-base, calc-base and calc-lib jsii models.

There is a few TODOs left as they are identified work required later on when implement calc

This is not supposed to build in the current status, as we still need to update the jsii-diff package.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

eladb commented:
Would be good to look at the generated result side-by-side with the current output. Initially I'd recommend writing the new generator so it will produce 1:1 output with the current generator. This way, we know we didn't regress. Then, we can start iterating.

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,614 @@ export default class Dotnet extends Target {
+
+        let implementsExpr = '';
+        if (cls.interfaces && cls.interfaces.length > 0) {
+            implementsExpr = ' implements ' + cls.interfaces.map(x => this.toNativeFqn(x));
```
This needs to be changed to : instead of implements

assyadh commented:
As discussed with @costleya , holding off the review as I have a few changes incoming for handling calc-base as well

assyadh commented:
Updated the generator to handle calc-base (introducing dependencies, multiple namespaces etc).

costleya commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+    private emitProperty(cls: spec.Type, prop: spec.Property, overrides: boolean = !!prop.overrides,
+                         emitInstanceGetter: boolean = false, datatype: boolean = false) {
+        const access = this.renderAccessLevel(prop);
+        const statc = prop.static ? 'static ' : '';
```
static

costleya commented on packages/jsii-pacmak/lib/targets/dotnet/filegenerator.ts:
```diff
@@ -0,0 +1,121 @@
+        propertyGroup.ele("GeneratePackageOnBuild", "true");
+        propertyGroup.ele("IncludeSymbols", "true");
+        propertyGroup.ele("IncludeSource", "true");
+        propertyGroup.ele("PackageVersion", this.assm.version);
```
put this.assm in a local var to avoid repeated lookup. Same with this.assm.targets.dotnet

assyadh replied:
Done, also refactored some other part

costleya commented on packages/jsii-pacmak/lib/targets/dotnet/nameutils.ts:
```diff
@@ -0,0 +1,83 @@
+
+/* We only want valid names for members */
+function isValidName(str: string) {
+    return str === null || str.match(/^\s*$/) !== null || str.indexOf('$') >= 0;
```
Rules: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/identifier-names

Missing rule here: Can only start with a letter or underscore.

Do we also want to proactively outlaw symbols like '='?

costleya commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        };
     }
 
     public async load(packageRoot: string) {
```
I see you are not adding return types to your functions. Please do that. It offers us better type-safety checks and intellisense.

assyadh commented:
Attaching the diffs for base and base-of-base generated files.

The only difference is that we now use the fully qualified name of property types instead of adding the property namespace to the using statement.

This will avoid running into bugs where we use the same type name twice, without prefixing the namespace.

[base-of-base-diff.txt](https://github.com/awslabs/jsii/files/3193006/base-of-base-diff.txt)
[base-diff.txt](https://github.com/awslabs/jsii/files/3193007/base-diff.txt)

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
-        // The DotNet generator does not currently support in-memory generation.
-        // Support can be added relatively easily if necessary.
+        // Override the openBlock to get a correct C# looking code block with the curly brace after the line
+        this.code.openBlock = function(s) {
```
renamed to text

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        super.generate(fingerprint);
+    }
+
+    public async save(outdir: string, tarball: string): Promise<any> {
```
Removed the promise altogether

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        // This is not in sync with the Old .NET generator where the name is scope-name-version.tgz.
+        // Hence we are saving the files ourselves here:
+        const assm = this.assembly;
+        const packageId: string = assm.targets!.dotnet!.packageId;
```
Done

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        for (const depName of Object.keys(assmDependencies)) {
+            const depInfo = assmDependencies[depName];
+            if (!this.dependencies.has(depName)) {
+                const namespace = depInfo.targets!.dotnet!.namespace;
```
Done

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        // For all base members
+        for (const base of interfaces) {
+            // Retrieve the interface name from the fqn
+            const baseFqn = base.substr(0, base.lastIndexOf('.'));
```
Done

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        }
+
+        this.code.line('get;');
+        if (!prop.immutable) {
```
This is on the onInterfaceProperty() method. We need it when generating the interface datatype, not here.

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        if (cls.interfaces && cls.interfaces.length > 0) {
+
+            // For all base members
+            for (const base of cls.interfaces) {
```
Done

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        // Nested classes will be dealt with during calc code generation
+        const nested = this.isNested(cls);
+        const inner = nested ? ' static' : '';
+        const absPrefix = abstract ? ' abstract' : '';
```
signature

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+
+        const className = nameutils.convertClassName(cls.name);
+
+        if (cls.docs!) {
```
failed copy pasta

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        /* tslint:disable-next-line no-unused-expression */
+        overrides;
+        const returnType = method.returns ? this.toDotNetType(method.returns.type) : 'void';
+        const statc = method.static ? 'static ' : '';
```
reserved name, renamed to staticKeyWord

costleya commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -338,38 +355,20 @@ class DotNetGenerator extends Generator {
-                    }
-                }
-            }
+            const implemented: [string[], string[]] = this.resolveImplementedInterfaces(cls);
```
const [baseTypeNamespaces, baseTypeNames] = this.resolveImplementedInterfaces(cls);

assyadh replied:
Done

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +128,677 @@ export default class Dotnet extends Target {
+        this.generateDependencyAnchorFile();
+
+        // Saving the generated code.
+        await this.code.save(outdir);
```
As discussed, not useful, creating a property or method won't absolve us from needing the this. statement.

assyadh commented:
Hi,

Since this has not yet been reviewed by a CDK team member, I have added some code to also handle the jsii-calc-lib model: virtual/override keywords, enums, generating documentation and slugifying some names.

costleya commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -376,23 +383,47 @@ class DotNetGenerator extends Generator {
+        if (initializers) {
+            if (initializers.parameters) {
+                jsiiAttribute = `[JsiiClass(nativeType: typeof(${className}), fullyQualifiedName: `
+                    + `"${cls.fqn}", parametersJson: "${JSON.stringify(cls.initializer!.parameters).replace(/"/g, '\\"')}")]`;
```
initializers.parameters

assyadh replied:
Good catch

costleya commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -376,23 +383,47 @@ class DotNetGenerator extends Generator {
+        }
         // Create the constructors:
+        let visibility = '';
         if (cls.abstract) {
```
const visibility = cls.abstract ? 'protected' : 'public';

assyadh replied:
Sure

costleya commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -451,54 +482,147 @@ class DotNetGenerator extends Generator {
+        if (overrides) {
+            // Add the override key word if the method is emitted for a proxy or data type or is defined on an ancestor
+            overrideKeyWord = 'override ';
+        } else {
```
"else if" to reduce nesting

assyadh replied:
Sure

costleya commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -451,54 +482,147 @@ class DotNetGenerator extends Generator {
+    // Emits all documentation depending on what is available in the jsii model
+    // Used by all kind of members + classes, interfaces, enums
+    private emitDocs(obj: spec.Method | spec.InterfaceType | spec.ClassType | spec.Property | spec.EnumType): void {
+        if (obj.docs) {
```
obj.docs in var

rix0rrr replied:
Early return maybe?

rix0rrr replied:
And yes, if you put it in a variable you can omit all the `!` annotations.

assyadh replied:
Done

costleya commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -564,23 +697,24 @@ class DotNetGenerator extends Generator {
     // This generates the body of the interface proxy or data type class
     // This loops through all the member and generates them
-    private emitInterfaceMembersForProxyOrDatatype(ifc: spec.InterfaceType | spec.ClassType, emitInstanceGetter: boolean, datatype: boolean): void {
+    private emitInterfaceMembersForProxyOrDatatype(ifc: spec.InterfaceType | spec.ClassType, datatype: boolean, proxy: boolean): void {
```
Are datatype and proxy flags mutually exclusive? If so, instead of using booleans, can we use an enum?

assyadh replied:
I taught about it, but chose not to because while they are mutually exclusive in the context of emitInterfaceMembersForProxyOrDatatype() these parameters are not in the context of emitProperty() (They can both be false if we are just emitting a property for a class).

rix0rrr commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +127,845 @@ export default class Dotnet extends Target {
+
+        }
+        const access = this.renderAccessLevel(method);
+        // TODO: handle async methods when impl calc-base and calc
```
`async` can be a JavaScript-only keyword. No need to translate into C#'s version of `async/await`, if that's what you were planning. But, that also depends on support from the runtime, and I'm not 100% sure on what that does today. Ask Aaron.

costleya replied:
The generator/runtime does not support async/await today. I think it was not prioritized since CDK users cannot use this pattern anyway, since everything has to be done in the constructor, which cannot be async.

We can revisit this for other non-cdk, jsii-based projects in the future. Any work should not require breaking changes.

rix0rrr commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +127,845 @@ export default class Dotnet extends Target {
+            }
+            if (obj.docs!.returns) {
+                this.code.line(`/// <returns>${obj.docs!.returns}</returns>`);
+            }
```
There's definitely some behavior that has disppeared here.

* Examples
* Parameter docstrings
* Arbitrary annotations (links, deprecated, etc.)

assyadh replied:
I have added the full documentation generation. Not all of it is required by the models currently generated but it should be there for the big *calc* model

assyadh commented:
> I would also like to see a run of the generator that introduces no (or minimal, and justifiable) diffs against vs the current generator produces.

[jsii-calc-base-diff.txt](https://github.com/awslabs/jsii/files/3228941/jsii-calc-base-diff.txt)
[jsii-calc-base-of-base-diff.txt](https://github.com/awslabs/jsii/files/3228942/jsii-calc-base-of-base-diff.txt)
[jsii-calc-lib-diff.txt](https://github.com/awslabs/jsii/files/3228943/jsii-calc-lib-diff.txt)


Here is a diff with the three currently supported models. The only noticeable diff is that we now use fully qualified type names in property and method return types (This avoid type collisions. This was not supported by the old generator and could lead to issues).

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +127,845 @@ export default class Dotnet extends Target {
-// ##################
-// # CODE GENERATOR #
-// ##################
+// #####################
```
Sure

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +127,845 @@ export default class Dotnet extends Target {
+    // Loops through the implemented interfaces and saves the type and namespace
+    // [0] The type names will be added to the class/ifc declaration
+    // [1] The namespace names will be added to the using statement
+    protected resolveImplementedInterfaces(ifc: spec.InterfaceType | spec.ClassType): [string[], string[]] {
```
Moved resolveDependencies and resolveImplementedInterfaces to private methods.

This moves the methods further down but it is OK.

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +127,845 @@ export default class Dotnet extends Target {
+        this.emitDocs(cls);
+        // Emit Jsii Attribute
+        let jsiiAttribute = `[JsiiClass(nativeType: typeof(${className}), fullyQualifiedName: "${cls.fqn}")]`;
+        const initializers = cls.initializer;
```
Good catch.

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +127,845 @@ export default class Dotnet extends Target {
+    }
+
+    // namespaces are handled implicitly by openFileIfNeeded().
+    protected onBeginNamespace(ns: string) {
```
The following would still be required right?

/* tslint:disable:no-empty */

assyadh commented on packages/jsii-pacmak/lib/targets/dotnet.ts:
```diff
@@ -126,44 +127,845 @@ export default class Dotnet extends Target {
+        let overrideKeyWord = '';
+        let virtualKeyWord = '';
+
+        const definedOnAncestor = this.isMemberDefinedOnAncestor(cls as spec.ClassType, method);
```
The method can be called for classes or interfaces.

Refactored the code to make it cleaner.

rix0rrr commented:
I don't understand why you need to attach the diff to the PR, and it's not showing up in the source? Maybe you need to run the full build with `UPDATE_DIFF=1`?

Also, the current build seems to have some issues:

```
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/AllTypes.cs(127,30): error CS1519: Invalid token ',' in class, struct, or interface member declaration [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/ClassWithDocs.cs(30,1): error CS1022: Type or namespace definition, or end-of-file expected [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/IDerivedStruct.cs(31,15): error CS1519: Invalid token '>' in class, struct, or interface member declaration [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/IDerivedStruct.cs(32,9): error CS1519: Invalid token '{' in class, struct, or interface member declaration [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/IDerivedStruct.cs(33,16): error CS1519: Invalid token ';' in class, struct, or interface member declaration [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/IDerivedStruct.cs(48,1): error CS1022: Type or namespace definition, or end-of-file expected [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Power.cs(9,80): error CS1001: Identifier expected [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Power.cs(9,80): error CS1003: Syntax error, ',' expected [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Statics.cs(33,29): error CS0145: A const field requires a value to be provided [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Statics.cs(34,74): error CS0145: A const field requires a value to be provided [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Statics.cs(35,29): error CS0145: A const field requires a value to be provided [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Statics.cs(36,49): error CS0145: A const field requires a value to be provided [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Build FAILED. 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/AllTypes.cs(127,30): error CS1519: Invalid token ',' in class, struct, or interface member declaration [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/ClassWithDocs.cs(30,1): error CS1022: Type or namespace definition, or end-of-file expected [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/IDerivedStruct.cs(31,15): error CS1519: Invalid token '>' in class, struct, or interface member declaration [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/IDerivedStruct.cs(32,9): error CS1519: Invalid token '{' in class, struct, or interface member declaration [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/IDerivedStruct.cs(33,16): error CS1519: Invalid token ';' in class, struct, or interface member declaration [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/IDerivedStruct.cs(48,1): error CS1022: Type or namespace definition, or end-of-file expected [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Power.cs(9,80): error CS1001: Identifier expected [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Power.cs(9,80): error CS1003: Syntax error, ',' expected [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Statics.cs(33,29): error CS0145: A const field requires a value to be provided [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Statics.cs(34,74): error CS0145: A const field requires a value to be provided [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Statics.cs(35,29): error CS0145: A const field requires a value to be provided [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
jsii-dotnet-runtime-test: Amazon/JSII/Tests/CalculatorNamespace/Statics.cs(36,49): error CS0145: A const field requires a value to be provided [/tmp/jsii-pacmak-codempIKUe/Amazon.JSII.Tests.CalculatorPackageId/Amazon.JSII.Tests.CalculatorPackageId.csproj] 
```

assyadh commented:
As discussed over Chime, Added an inline diff.

I also added a few changes to the build scripts to neutralize the build of calc. This way we have the branch buildable in the current state.
